### PR TITLE
[Storage] Add Structured Message CRC64 content validation for blob upload and download

### PR DESCRIPTION
**Pull request by tanyasethi-msft**
This PR includes changes from already approved PRs. 
  
## Summary of this PR 

  - Adds end-to-end support for **Structured Message (SM) CRC64** content validation across all blob types (block, append, page), for both upload and download operations
  - Introduces `TransferValidationTypeComputeStructuredMessageCRC64` — a new `TransferValidationType` that computes per-segment CRC64 checksums using the structured message binary format, providing stronger
  data integrity guarantees than whole-blob MD5/CRC64
  - Implements a streaming SM encoder/decoder in `internal/shared/` that wraps request/response bodies transparently — no changes needed to existing upload/download call patterns

  ### What's included

  1. **Core library** (`internal/shared/structured_message.go`) — SM binary format encoder & decoder with per-segment CRC64 computation and validation
  2. **Upload integration** — SM CRC64 support on `blockblob.Upload`, `UploadBuffer`, `UploadStream`, `UploadFile`, `StageBlock`, `CommitBlockList`, `appendblob.AppendBlock`, and `pageblob.UploadPages`
  3. **Download integration** — SM CRC64 validation on `blob.DownloadStream`, `DownloadBuffer`, and `DownloadFile` with automatic structured message decoding and per-segment integrity checks
  4. **Public API** — `blob.TransferValidationTypeComputeStructuredMessageCRC64(segmentSize)` (use 0 for default 4 MB segments)
  5. **Tests** — Unit tests for encoder/decoder, Integration tests across all blob types covering round-trips, edge cases (empty blob, single byte, exact segment boundary, multi-segment), and error paths
  6. **Benchmarks** — encoder/decoder benchmarks in `internal/shared/` and upload benchmarks in `blockblob/`
  7. **Examples** — `Example_blockblob_UploadWithContentValidation` and `Example_blob_DownloadWithContentValidation`

  ### Usage

  ```go
  // Upload with SM CRC64 (0 = default 4 MB segment size)
  _, err = blockBlobClient.Upload(ctx, body, &blockblob.UploadOptions{
      TransactionalValidation: blob.TransferValidationTypeComputeStructuredMessageCRC64(0),
  })

  // Download with SM CRC64 validation
  resp, err := blobClient.DownloadStream(ctx, &blob.DownloadStreamOptions{
      TransactionalValidation: blob.TransferValidationTypeComputeStructuredMessageCRC64(0),
  })
```
  Test plan

  - unit tests for SM encoder/decoder (internal/shared/structured_message_test.go)
  - integration tests across block blob, append blob, page blob, and blob download
  - Edge cases: empty blob, single byte, exact 4 MB segment boundary, multi-block round-trips
  - Benchmarks for encoder/decoder and upload paths

- [x] The purpose of this PR is explained in this or a referenced issue.
- [x] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [x] Tests are included and/or updated for code changes.
- [x] Updates to module CHANGELOG.md are included.
- [x] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go

### CODE REVIEW

File: sdk/storage/azblob/blob/client.go [hunk 1]
```diff
@@ -389,7 +390,11 @@ func (b *Client) downloadBuffer(ctx context.Context, writer io.WriterAt, o downl
 				return err
 			}
 			if computeReadLength {
-				atomic.AddInt64(&dataDownloaded, *dr.ContentLength)
+				if dr.ContentRange != nil {
+					atomic.AddInt64(&dataDownloaded, parseContentRangeLength(*dr.ContentRange))
+				} else {
+					atomic.AddInt64(&dataDownloaded, *dr.ContentLength)
+				}
 			}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azblob/blob/client.go:400]: made changes.

File: sdk/storage/azblob/internal/exported/transfer_validation_option.go [hunk 2]
```diff
@@ -55,6 +55,45 @@ func (c TransferValidationTypeMD5) Apply(rsc io.ReadSeekCloser, cfg generated.Tr
 
 func (TransferValidationTypeMD5) notPubliclyImplementable() {}
 
+// TransferValidationTypeComputeStructuredMessageCRC64 is a TransferValidationType that computes
+// per-segment CRC64 checksums using the structured message binary format.
+// The body is wrapped in a streaming SMEncoder that produces SM-encoded output on Read().
+// segmentSize specifies the maximum segment size in bytes (use 0 for default 4MB).
+func TransferValidationTypeComputeStructuredMessageCRC64(segmentSize int) TransferValidationType {
+	return &transferValidationTypeSMCRC64{segmentSize: segmentSize}
+}
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azblob/internal/exported/transfer_validation_option.go:64]: updated the doc

File: sdk/storage/azblob/CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 ## 1.8.0-beta.2 (Unreleased)
 
 ### Features Added
+* Added support for Structured Message CRC64 content validation on upload and download operations using `TransferValidationTypeComputeStructuredMessageCRC64`.
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azblob/CHANGELOG.md:6]: The CHANGELOG entry says "upload and download operations" — that covers all download variants.

File: sdk/storage/azblob/internal/shared/structured_message.go [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+
+	SMDefaultSegmentSize = 4 * 1024 * 1024 // 4MB
+
+	SMHeaderSize         = 13 // version(1) + msgLen(8) + flags(2) + numSegments(2)
+	SMSegmentHeaderSize  = 10 // segNum(2) + dataLen(8)
+	SMSegmentFooterSize  = 8  // CRC64
+	SMMessageTrailerSize = 8  // CRC64
+
+	SMHeaderValue = "XSM/1.0; properties=crc64"
+)
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azblob/internal/shared/structured_message.go:28]: they're used across internal subpackages (internal/exported/ imports internal/shared/), so they must be
  exported. No change needed.

File: sdk/storage/azblob/internal/shared/structured_message_test.go [hunk 5]
```diff
@@ -0,0 +1,986 @@
...
+}
+
+func TestDecoderInvalidBody(t *testing.T) {
+	body := io.NopCloser(bytes.NewReader([]byte{0xFF, 0x01, 0x02}))
+	dec := NewSMDecoder(body)
+
+	_, err := io.ReadAll(dec)
+	require.Error(t, err)
+}
+
```
[the remark this answers is not in the document]
Comment (tanyasethi-msft) [reply sdk/storage/azblob/internal/shared/structured_message_test.go:350]: Fair suggestion for robustness, but low priority for this PR. Could be a follow up

File: sdk/storage/azblob/blockblob/client.go [hunk 6]
```diff
@@ -527,8 +536,11 @@ func (bb *Client) UploadBuffer(ctx context.Context, buffer []byte, o *UploadBuff
 		uploadOptions = *o
 	}
 
-	// If user attempts to pass in their own checksum, errors out.
-	if uploadOptions.TransactionalValidation != nil && reflect.TypeOf(uploadOptions.TransactionalValidation).Kind() != reflect.Func {
+	// If user attempts to pass in their own pre-computed checksum, errors out.
+	// Structured message CRC64 is allowed because it computes per-block checksums on-the-fly.
+	if uploadOptions.TransactionalValidation != nil &&
```
Comment (jhendrixMSFT) [inline sdk/storage/azblob/blockblob/client.go:541]: I'm a little unclear on what we're trying to validate/prevent here.  Since `TransactionalValidation` can't be implemented outside of `azblob`, it seems unnecessary.  I get that we're expanding on existing code, but is this really necessary?

**Review by jhendrixMSFT — APPROVED** [review]

File: sdk/storage/azblob/blockblob/client.go (hunk 6 above)
Comment (tanyasethi-msft) [reply sdk/storage/azblob/blockblob/client.go:541]: The only values that canreach here are SDK-provided types.
The existing reflect.TypeOf(...).Kind() != reflect.Func check (pre-dating this PR) was already a workaround to distinguish pre-computed checksums (TransferValidationTypeCRC64, TransferValidationTypeMD5 — which are type aliases, not funcs) from compute-on-the-fly types (TransferValidationTypeComputeCRC64 — which is a transferValidationTypeFn). The intent is to reject pre-computed checksums on multi-block uploads since they can't apply a whole-blob checksum per-block.

This PR adds exported.GetStructuredBodyType(...) == "" to also allow SM CRC64 through, since it computes per-block checksums on-the-fly (same category as ComputeCRC64).

I agree the reflection-based check is fragile. A cleaner approach would be to add a method to the TransferValidationType interface (e.g., supportsMultiBlockUpload() bool) instead of using reflect + type-checking. But that's a refactor of the existing pattern — happy to do it in this PR or as a follow-up. What do you prefer?

Comment (jhendrixMSFT) [reply sdk/storage/azblob/blockblob/client.go:541]: Got it, thanks for the detailed explanation.

Agreed it would be good to move away from the reflection-based check.  I opened https://github.com/Azure/azure-sdk-for-go/issues/26791 to track that.
